### PR TITLE
raftstore: reduce the frequency of flushing raft messages

### LINKED ISSUE #11309: Reduce the frequency of flushing raft message
## Development Task

https://github.com/tikv/tikv/pull/11111 introduced a config `raft-msg-flush-interval-us` which is used to reduce the frequency of flushing raft messages. 
However, after some investigation, I found many flush events come from the `pause` function when the store threads are not full.

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Issue Number: close #11309

Problem Summary:
See #11309.

### What is changed and how it works?

What's Changed:
If there is no active fsm in store thread, it will wait an additional `raft_msg_flush_interval_us`, if there is still no active fsm, the `pause` function should be called.

This PR also changes some other small parts
1. change the default `store-io-pool-size` to 1 because if it's basically enough for other default configs (e.g. 2 store threads, 2 apply threads and 5 grpc threads) in most cases. Users can increase this config and other relevant configs to achieve better performance if the CPU is sufficient.
2. remove the config `cmd-batch-concurrent-ready-max-count` from `config-template.toml` because this config should not be changed in almost all cases. (BTW, the config name may be changed in future)
3. `raft_entry_max_size` can be updated dynamically. But the `BatchRaftCmdRequestBuilder` always uses the original one.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- No.

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/raftstore/src/store/config.rs [hunk 1]
```diff
@@ -299,7 +299,7 @@ impl Default for Config {
             local_read_batch_size: 1024,
             apply_batch_system: BatchSystemConfig::default(),
             store_batch_system: BatchSystemConfig::default(),
-            store_io_pool_size: 2,
+            store_io_pool_size: 1,
```
Comment (BusyJay) [inline components/raftstore/src/store/config.rs:302]: Why?

File: components/batch-system/src/config.rs [hunk 2]
```diff
@@ -11,6 +11,7 @@ pub struct Config {
     pub pool_size: usize,
     pub reschedule_duration: ReadableDuration,
     pub low_priority_pool_size: usize,
+    pub before_pause_wait_us: Option<u64>,
```
Comment (BusyJay) [inline components/batch-system/src/config.rs:14]: How about reusing `raft_msg_flush_interval_us`?

File: components/raftstore/src/store/config.rs (hunk 1 above)
Comment (gengliqi) [reply components/raftstore/src/store/config.rs:302]: You can see the description.

File: components/batch-system/src/config.rs (hunk 2 above)
Comment (gengliqi) [reply components/batch-system/src/config.rs:14]: This config is useless for apply batch system. And the `raft_msg_flush_interval_us` belongs to raftstore not batch system so I think just using two configs is ok.
The default value of store batch system's `before_pause_wait_us` is `raft_msg_flush_interval_us `.

Comment (BusyJay) [reply components/batch-system/src/config.rs:14]: You can mark it skip serde skip and use `raft_msg_flush_interval_us` to initialize the value.

File: components/batch-system/src/config.rs [hunk 3]
```diff
@@ -11,6 +11,9 @@ pub struct Config {
     pub pool_size: usize,
     pub reschedule_duration: ReadableDuration,
     pub low_priority_pool_size: usize,
+    #[doc(hidden)]
+    #[serde(skip_serializing)]
```
Comment (BusyJay) [inline components/batch-system/src/config.rs:15]: It should be `#[serde(skip)]`.

File: components/batch-system/src/batch.rs [hunk 4]
```diff
@@ -273,9 +274,25 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
...
-            if let Ok(fsm) = self.fsm_receiver.recv() {
-                return batch.push(fsm);
+            if let Some(d) = self.before_pause_wait_us {
+                channel::select! {
+                    recv(self.fsm_receiver) -> msg => {
+                        if let Ok(fsm) = msg {
+                            return batch.push(fsm);
+                        }
+                    }
+                    recv(tick(Duration::from_micros(d))) -> _ => {
```
Comment (BusyJay) [inline components/batch-system/src/batch.rs:284]: Why not use `after`?

**Review by BusyJay — COMMENTED** [review]
Can you also post the improvement in PR?

Rest LGTM

**Comment by gengliqi** [comment]
From left to right, master, this PR with 250us `raft_msg_flush_interval_us `, this PR with 500us `raft_msg_flush_interval_us `
gsMmZdoUJR

File: components/raftstore/src/store/config.rs [hunk 5]
```diff
@@ -489,6 +489,8 @@ impl Config {
         } else {
             self.store_batch_system.max_batch_size = Some(1024);
         }
+        self.store_batch_system.before_pause_wait_us =
+            Some(std::cmp::min(self.raft_msg_flush_interval_us, 1000));
```
Comment (BusyJay) [inline components/raftstore/src/store/config.rs:493]: I suggest to change the time to `ReadableDuration` for consistency.

**Review by BusyJay — APPROVED** [review]
Rest LGTM

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Pull request merged**